llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -0,0 +1,257 @@
+    if (F.getCallingConv() == CallingConv::SPIR_KERNEL &&
+        F.hasFnAttribute("sycl-module-id")) {
+      auto Id = F.getFnAttribute("sycl-module-id");
+      auto Val = Id.getValueAsString();
+      ResKernelsSet[Val].push_back(&F);
+    }
+  }
+}
+
+static void
+splitModule(Module &M,
+            std::map<std::string, std::vector<Function *>> &KernelsSet,
+            std::vector<std::unique_ptr<Module>> &ResModules,
+            std::vector<std::string> &ResSymbolsLists) {
+  for (auto &It : KernelsSet) {
+    // For each group of kernels collect all dependencies.
+    SetVector<GlobalValue *> GVs;
+    std::vector<llvm::Function *> Workqueue;
+    std::string SymbolsList;
+
+    for (auto &F : It.second) {
+      GVs.insert(F);
+      Workqueue.push_back(F);
+      SymbolsList =
+          (Twine(SymbolsList) + Twine(F->getName()) + Twine("\n")).str();
+    }
+
+    while (!Workqueue.empty()) {
+      Function *F = &*Workqueue.back();
+      Workqueue.pop_back();
+      for (auto &BB : *F) {
+        for (auto &I : BB) {
+          if (CallBase *CB = dyn_cast<CallBase>(&I)) {
+            if (Function *CF = CB->getCalledFunction()) {
+              if (!CF->isDeclaration() && !GVs.count(CF)) {
+                GVs.insert(CF);
+                Workqueue.push_back(CF);
+              }
+            }
+          }
+        }
+      }
+    }
+
+    // It's not easy to trace global variable's uses inside needed functions
+    // because global variable can be used inside a combination of operators, so
+    // mark all global variables as needed and remove dead ones after
+    // extraction.
+    for (auto &G : M.globals()) {
+      GVs.insert(&G);
+    }
+
+    // Clone the module, understand which globals we need to extract from the
+    // clone.
+    ValueToValueMapTy VMap;
+    std::unique_ptr<Module> MClone = CloneModule(M, VMap);
+    std::vector<GlobalValue *> GVsInClone(GVs.size());
+    int I = 0;
+    for (GlobalValue *GV : GVs) {
+      GVsInClone[I] = cast<GlobalValue>(VMap[GV]);
+      ++I;
+    }
+
+    // TODO: Use the new PassManager instead?
+    legacy::PassManager Extract;
+
+    // Extract needed globals.
+    Extract.add(createGVExtractionPass(GVsInClone, /* deleteS */ false));
+    Extract.run(*MClone.get());
+
+    // Extactor pass sets external linkage to all globals. Return linkage back.
+    for (auto &G : MClone->globals()) {
+      if (G.getVisibility() == GlobalValue::HiddenVisibility) {
+        G.setLinkage(GlobalValue::InternalLinkage);
+      }
+    }
+
+    legacy::PassManager Passes;
+    // Do cleanup.
+    Passes.add(createGlobalDCEPass());           // Delete unreachable globals.
+    Passes.add(createStripDeadDebugInfoPass());  // Remove dead debug info.
+    Passes.add(createStripDeadPrototypesPass()); // Remove dead func decls.
+    Passes.run(*MClone.get());
+
+    // Save results.
+    ResModules.push_back(std::move(MClone));
+    ResSymbolsLists.push_back(SymbolsList);
```

Comment:
I used std::move because Cloning function returns `std::unique_ptr` and I needed to put this module object to resulting collection of modules. I think `std::move` also can be applied to `SymbolsList` as an optimization.

Reply:
ok, I see